### PR TITLE
Implement profile page

### LINKED ISSUE #29: Create User Profile Page
Implement public profile page at `/u/[handle]`

### LINKED ISSUE #33: Show Recently Attended Events On User Profiles
Cannot complete until https://github.com/acmucsd/membership-portal/issues/347 is closed

### PR DESCRIPTION
**Pull request by raymosun**
# Info

Closes #29 

# Description

Implements the user profile page.

Note: the original profile page branch died and wouldn't deploy, possibly because it's too old or something 🤷 

## Changes

- `/profile` redirects to `/u/[user_handle]`
- Creates the `UserProfilePage` component which renders a user profile page
- Creates the `HandleNotFound` component which tells users a handle does not exist
- Creates a new page at `/u/[handle]` which retrieves the current user, the handle's user, and their attendances, then renders the correct component 
- Updates the `utils.js` functions `getUserRank()`  and `hashUser()` with their real functionality
- Creates the `GifSafeImage` component which automatically makes next/image Images unoptimized when `src` is a gif
- Side effect: deleted `width` and `height` attributes from some svgs so they can be used at different sizes on this PR; existing usages have (hopefully) all been updated to manually  set the size
- `/constants/majors.json` refactored to a `.ts` file

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) 

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [x] locally on mobile - use https://ngrok.io to get a copy on a mobile device
- [x] on the live deployment preview on Desktop.
- [x] on the live deployment preview on Mobile.
- [ ] I have run and passed all new and existing Cypress tests. Add screenshots below.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented my code's `src/lib` functions and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots

desktop- own profile - light mode
image

desktop- other's profile - light mode
image

mobile - other's profile - dark mode
image

handle not found - dark mode
image

### CODE REVIEW

**Comment by farisashai** [comment]
I would remove the edit button in the about section since it goes to the same place

**Comment by farisashai** [comment]
image

This incentive seems personal so it's strange to see it on someone else's profile. I think we should hide it if they're not viewing their own

Can you also add more gap above the text and left align it

**Comment by farisashai** [comment]
image

More gap between the rank and points

Also specify "x **All-Time** Leaderboard Points"

**Comment by farisashai** [comment]
image

Hide the div for social media links if it's empty

**Comment by farisashai** [comment]
Fun challenge for later: customizing the pink background per user so everyone's looks different  

my idea was that you take their past attendances and proportionally use each of the community colors that they're engaged with in some kind of randomly generated blobs or patterns so the background represents them and it could change every time they attend a new event

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx]:
```suggestion
import { EventCard } from '@/components/events';
```

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx]: Just confirming does this support gifs correctly?

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx]: maybe clicking on this can copy the link to their profile to the clipboard and notify the user in a toast

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx:79]: Do you think an edit pencil or a settings icon looks better here?

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx:157]: I think that even though the user has permission to view their own attendances here, it might be confusing if they have it set to private and they think this page is an accurate reflection of how other people are seeing their profile. 

Options are either hide it if their attendances are private or add some border/colo/text indicating that this is hidden for other users.

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx:156]: Isn't there an event carousel specific component used on the home page?

File: src/components/profile/index.ts
Comment (farisashai) [inline src/components/profile/index.ts]: can you match the syntax of the other exports

File: src/lib/utils.ts
Comment (farisashai) [inline src/lib/utils.ts]: can we move the list of ranks to the constants folder and import from there

File: src/pages/profile.tsx
Comment (farisashai) [inline src/pages/profile.tsx]: cost UserProfilePage = () => null

Let's remove the body completely for clarity since it's never being used

File: src/pages/u/[handle].tsx
Comment (farisashai) [inline src/pages/u/[handle].tsx]: weird detail but can we rename this so the signedInUser variable is called user and the other one has a different name when we pass the props. the way the repo is structured, the user prop always refers to the logged in user on all other pages so it'll break in my open PR if we don't follow that convention

File: src/pages/u/[handle].tsx
Comment (farisashai) [inline src/pages/u/[handle].tsx]: Why do we return this if the profile we're viewing has private attendance?

Comment (farisashai) [reply src/pages/u/[handle].tsx]: I think this can be simplified a little so we have one return block and some of the variables may be undefined but it's not a big deal

**Review by farisashai — CHANGES_REQUESTED** [review]
left some visual nits as comments in addition to these code comments. code looks excellent very clean and readable and the page also looks great excellent work 🤩

**Comment by raymosun** [comment]
> image
> 
> This incentive seems personal so it's strange to see it on someone else's profile. I think we should hide it if they're not viewing their own
> 
> Can you also add more gap above the text and left align it

Have you seen the version on other profiles that uses their name instead of "My Progress"? I think it doesn't feel too weird and helps the page not feel so empty.
image

File: src/components/profile/UserProfilePage/index.tsx
Comment (raymosun) [reply src/components/profile/UserProfilePage/index.tsx]: Thanks for the catch, implemented a new `GifSafeImage` component and refactored other usages of `isSrcAGif()`

**Comment by raymosun** [comment]
> I would remove the edit button in the about section since it goes to the same place

Messing with having the the second edit button link to directly editing the 'About me' section, thoughts?

**Comment by farisashai** [comment]
> > image
> 
> > 
> 
> > This incentive seems personal so it's strange to see it on someone else's profile. I think we should hide it if they're not viewing their own
> 
> > 
> 
> > Can you also add more gap above the text and left align it
> 
> 
> 
> Have you seen the version on other profiles that uses their name instead of "My Progress"? I think it doesn't feel too weird and helps the page not feel so empty.
> 
> image
> 
> 
> 
> 

this is probably fine to leave

**Comment by farisashai** [comment]
> > I would remove the edit button in the about section since it goes to the same place
> 
> 
> 
> Messing with having the the second edit button link to directly editing the 'About me' section, thoughts?

I think it's simpler to just remove the button since they go to the same page and navigation from there is very easy to the about section

**Comment by raymosun** [comment]
> > > I would remove the edit button in the about section since it goes to the same place
> > 
> > 
> > Messing with having the the second edit button link to directly editing the 'About me' section, thoughts?
> 
> I think it's simpler to just remove the button since they go to the same page and navigation from there is very easy to the about section

Maybe we need to move the location of the edit button entirely, it being inside one of the cards kinda makes it look like you're editing exactly the stuff in the card.

P.S. why can't i reply in a thread like a code comment 😭

File: src/pages/u/[handle].tsx
Comment (raymosun) [reply src/pages/u/[handle].tsx]: Refactored this in latest commit 👍 https://github.com/acmucsd/membership-portal-ui-v2/pull/117/commits/461374f7a5a46b56c871b719e7bbb552783881d3

File: src/pages/u/[handle].tsx
Comment (raymosun) [reply src/pages/u/[handle].tsx]: Implemented in this in latest commit 👍 https://github.com/acmucsd/membership-portal-ui-v2/pull/117/commits/461374f7a5a46b56c871b719e7bbb552783881d3

File: src/pages/profile.tsx
Comment (raymosun) [reply src/pages/profile.tsx]: Implemented in this in latest commit 👍 https://github.com/acmucsd/membership-portal-ui-v2/pull/117/commits/461374f7a5a46b56c871b719e7bbb552783881d3

File: src/lib/utils.ts
Comment (raymosun) [reply src/lib/utils.ts]: Implemented in this in latest commit 👍 https://github.com/acmucsd/membership-portal-ui-v2/pull/117/commits/461374f7a5a46b56c871b719e7bbb552783881d3

File: src/components/profile/index.ts
Comment (raymosun) [reply src/components/profile/index.ts]: Implemented in this in latest commit 👍 https://github.com/acmucsd/membership-portal-ui-v2/pull/117/commits/461374f7a5a46b56c871b719e7bbb552783881d3

File: src/components/profile/UserProfilePage/index.tsx
Comment (raymosun) [reply src/components/profile/UserProfilePage/index.tsx:157]: Added a note in latest commit 👍 image

File: src/components/profile/UserProfilePage/index.tsx
Comment (raymosun) [reply src/components/profile/UserProfilePage/index.tsx]: Implemented in this in latest commit 👍 https://github.com/acmucsd/membership-portal-ui-v2/pull/117/commits/461374f7a5a46b56c871b719e7bbb552783881d3

File: src/components/profile/UserProfilePage/index.tsx
Comment (raymosun) [reply src/components/profile/UserProfilePage/index.tsx:79]: Hm i quite like the edit pencil if it's in the current place on the profile card, but maybe the lack of a cog anywhere will confuse users about where to find settings like resume upload and email.

File: src/components/common/GifSafeImage/index.tsx
Comment (farisashai) [inline src/components/common/GifSafeImage/index.tsx:9]: you don't need to destructure the alt prop since you're not using it and it will be included in restProps automatically but it's fine either way

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx]: You can move that trailing space inside both template literal strings

File: src/lib/constants/ranks.json
Comment (farisashai) [inline src/lib/constants/ranks.json]: it might be better to store this array in a typescript (.ts) file and export from there so it can infer types

File: src/lib/utils.ts
Comment (farisashai) [inline src/lib/utils.ts]: import this as a type if that's the only usage so it gets removed automatically when the build is compiled down to JavaScript

File: src/pages/profile.tsx
Comment (farisashai) [inline src/pages/profile.tsx]: I think this should actually be a fresh data fetch because unlike their uuid, the user can change their own handle easily, breaking this link

**Review by farisashai — APPROVED** [review]
everything looks great minus some small tweaks - I would wait and try to get one more review to take a look before merging in case anyone has thoughts

File: src/components/common/GifSafeImage/index.tsx
Comment (raymosun) [reply src/components/common/GifSafeImage/index.tsx:9]: I tried without it but there's a lint rule requiring `alt` prop explicitly on Image components and it's not smart enough to figure out it must be there

File: src/components/profile/UserProfilePage/index.tsx
Comment (raymosun) [reply src/components/profile/UserProfilePage/index.tsx]: LOL that's so dumb 😆 I think the formatter put it there, thanks for catch

File: src/pages/profile.tsx
Comment (raymosun) [reply src/pages/profile.tsx]: So true, implemented in latest commit https://github.com/acmucsd/membership-portal-ui-v2/pull/117/commits/27bbe33e3309ee40df265b0da388ac43afd7609c

File: src/lib/constants/ranks.json
Comment (raymosun) [reply src/lib/constants/ranks.json]: I went ahead and updated majors.json to be like this too 👍

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx:53]: this text should be center aligned

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx]: the mail link href needs to be pre-formatted from EMAIL to mailto:EMAIL - it doesnt work right now. the mail link from the edit profile link is formatted correctly

File: src/pages/profile.tsx
Comment (farisashai) [inline src/pages/profile.tsx:14]: when we fetch the user cookie, can we update the user cookie just as a sort of checkpoint that the data is as fresh as possible in case the access type or something weird updates

**Review by farisashai — APPROVED** [review]
i finally looked at this on desktop - we can just merge after these small changes

**Comment by farisashai** [comment]
if you try to view someone's profile where they set attendance to private, the recently attended events box still shows the title with the cards missing - you should hide it completely or leave a message that the attendance is private

File: public/assets/icons/facebook-icon.svg
Comment (SheepTester) [inline public/assets/icons/facebook-icon.svg:1]: the edit profile page uses react-icons, mostly out of laziness (didn't feel like uploading images). we should probably use one of them for consistency

File: src/components/profile/HandleNotFound/index.tsx
Comment (SheepTester) [inline src/components/profile/HandleNotFound/index.tsx:1]: Wondering if it'd be a good idea to name this to `UserHandleNotFound` (ie dont start with "handle"), since I initially thought this was just a generic handler for something not found

File: src/components/profile/Preview/style.module.scss
Comment (SheepTester) [inline src/components/profile/Preview/style.module.scss]: One issue with this is that if the bio is long and the screen is short, the user has to scroll to the bottom of the page to scroll to the bottom of the preview. Is this fine?

File: src/components/profile/UserProfilePage/index.tsx
Comment (SheepTester) [inline src/components/profile/UserProfilePage/index.tsx]: You can use the `SocialMediaIcon` component, used by the edit profile page

Comment (alexzhang1618) [inline src/components/profile/UserProfilePage/index.tsx]: can we use the component SocialMediaIcon from Sean's PR here instead?

File: src/components/profile/UserProfilePage/index.tsx
Comment (SheepTester) [inline src/components/profile/UserProfilePage/index.tsx]: Might the code be cleaner if you iterated over `handleUser.userSocialMedia` instead of `socialMediaIcons`? I think the code for this in this and the `Preview` component should be the same, at the very least

File: src/lib/api/EventAPI.ts
Comment (alexzhang1618) [inline src/lib/api/EventAPI.ts]: Do we still need this if we can use `getAttendancesforCurrentUser` instead?

File: src/lib/utils.ts
Comment (alexzhang1618) [inline src/lib/utils.ts:121]: Is there a need to change this from taking `PublicProfile` to `points`? It seems more intuitive to `getUserRank` by passing in a user instead of a number

File: src/components/profile/UserProfilePage/style.module.scss
Comment (SheepTester) [inline src/components/profile/UserProfilePage/style.module.scss:189]: (This is also an issue with the `Preview` component) It appears newlines in the `bio` get removed. You could add

```suggestion
    .bioSection {
      white-space: pre-wrap;
```

so the text shows the same as how it appears in the text input (keeps line breaks, duplicate spaces, but also allows text wrapping)

Comment (SheepTester) [reply src/components/profile/UserProfilePage/style.module.scss:189]: you should also add `word-break: break-word;` or whatever it's called, so

image image

**Review by alexzhang1618 — COMMENTED** [review]
Some visual comments: 
Are these icons centered? They seem a little off
Screenshot 2024-01-09 at 1 09 49 PM

Same here with the trophy icon: 
- do we need to specify `All Time Leaderboard Points`? It seems like it makes some of the wrapping weird
- the edit button icon seems a little close to the content
Screenshot 2024-01-09 at 1 10 13 PM

This carousel might get weird if the user has a ton of attended events or if it's mobile:
Screenshot 2024-01-09 at 1 11 33 PM
Screenshot 2024-01-09 at 1 12 29 PM

File: src/components/profile/UserProfilePage/index.tsx
Comment (SheepTester) [inline src/components/profile/UserProfilePage/index.tsx:135]: Maybe you could just hide the entire Bio section if the user bio is empty

File: src/components/profile/UserProfilePage/index.tsx
Comment (SheepTester) [inline src/components/profile/UserProfilePage/index.tsx]: "All-Time Leaderboard Points" feels kind of verbose. Maybe just "points" (since the points for this year aren't shown anywhere on the profile, I think) or "total points"?

File: src/components/profile/UserProfilePage/style.module.scss
Comment (SheepTester) [inline src/components/profile/UserProfilePage/style.module.scss]: you should also center the handle

image

File: src/lib/utils.ts
Comment (SheepTester) [inline src/lib/utils.ts:120]: is the cast to string necessary?

**Review by SheepTester — APPROVED** [review]
looks pretty good! just some nitpicks

Also, since the bottom of the page is just the event attendance history, I think it'd be worth changing it to a grid (kind of like Instagram) rather than just a scroller, because it's easier to scroll down than to the side, and on Android, the scrollbar ends up looking really tiny

image

the contrast between the blue and grey, especially on light mode, is kind of weird. you could make the grey lighter, or add a gap

image


image

**Comment by raymosun** [comment]
> if you try to view someone's profile where they set attendance to private, the recently attended events box still shows the title with the cards missing - you should hide it completely or leave a message that the attendance is private

weird, i can't seem to reproduce this, can you give more details? 

image

File: src/lib/utils.ts
Comment (raymosun) [reply src/lib/utils.ts:121]: hm slightly, but i'm currently getting the next rank by passing `user.points + 100`, so i think this is okay for now

File: src/lib/utils.ts
Comment (raymosun) [reply src/lib/utils.ts:120]: apparently, yes unfortunately :(

**Comment by raymosun** [comment]
Implemented some changes in latest commit https://github.com/acmucsd/membership-portal-ui-v2/pull/117/commits/8a0198fd280f38cb33d2fb9007c79be99804e06c but I want to update icons also soon

File: src/components/profile/UserProfilePage/style.module.scss
Comment (raymosun) [reply src/components/profile/UserProfilePage/style.module.scss]: Implemented in latest commit https://github.com/acmucsd/membership-portal-ui-v2/commit/8a0198fd280f38cb33d2fb9007c79be99804e06c 👍

File: src/components/profile/UserProfilePage/index.tsx
Comment (raymosun) [reply src/components/profile/UserProfilePage/index.tsx]: Yeah I agree, implemented in latest commit https://github.com/acmucsd/membership-portal-ui-v2/commit/8a0198fd280f38cb33d2fb9007c79be99804e06c 👍

File: src/components/profile/UserProfilePage/index.tsx
Comment (raymosun) [reply src/components/profile/UserProfilePage/index.tsx:135]: I think the card looks kinda empty and weird without it so I'm going to leave it for now

File: src/components/profile/UserProfilePage/style.module.scss
Comment (raymosun) [reply src/components/profile/UserProfilePage/style.module.scss:189]: Good catch! Implemented in latest commit https://github.com/acmucsd/membership-portal-ui-v2/commit/8a0198fd280f38cb33d2fb9007c79be99804e06c 👍

File: src/lib/utils.ts
Comment (SheepTester) [reply src/lib/utils.ts:120]: You could also do `ranks[index] ?? ''`, defaulting to some string, instead of casting

File: src/components/profile/Preview/index.tsx
Comment (farisashai) [inline src/components/profile/Preview/index.tsx]:
```suggestion
import { Typography, GifSafeImage  } from '@/components/common';
```

File: src/pages/events.tsx
Comment (farisashai) [inline src/pages/events.tsx]: these two api calls should be batched with a Promise.all

File: src/pages/u/[handle].tsx
Comment (farisashai) [inline src/pages/u/[handle].tsx]: not a fan of this syntax, can we move any `await` calls to their own variable

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx:59]: pointer: cursor;

File: src/pages/u/[handle].tsx
Comment (raymosun) [reply src/pages/u/[handle].tsx]: i did it to avoid unnecessary api calls and `let`, would you prefer if the null was the second option maybe?

File: cypress/e2e/pages/edit-profile.cy.ts
Comment (farisashai) [inline cypress/e2e/pages/edit-profile.cy.ts]:
```suggestion
describe('Edit Profile Page', () => {
```

File: cypress/support/commands.ts
Comment (farisashai) [inline cypress/support/commands.ts:34]: does it work if you use the typeInForm command directly here instead?

File: src/pages/u/[handle].tsx
Comment (farisashai) [reply src/pages/u/[handle].tsx]: you could just wrap it in an if statement but honestly not a big deal to leave it either

```
let attendances = null;
if condition {
    attendances = await ...
}
```

File: src/components/profile/UserProfilePage/index.tsx
Comment (farisashai) [inline src/components/profile/UserProfilePage/index.tsx]: display: flex;
gap: 4px;

the icons are touching each other right now

**Review by farisashai — APPROVED** [review]
couple of tiny nits but amazing work overall 🎉 code quality is very clean and thoughtful and the UI is in great shape

**Comment by farisashai** [comment]
Closes #33

File: cypress/support/commands.ts
Comment (raymosun) [reply cypress/support/commands.ts:34]: it doesn't work with the current implementation because the login page labels in the placeholder instead of with a <label>, the typeInForm command could be made more general in a future pr maybe

**Pull request merged**
